### ISSUE TITLE
Which Model I can use ? (Pytorch, Tensorflow , ONNX) 

### ISSUE THREAD

**Issue by akshin18** [gomlx/gomlx#80]
Could you please clarify for me which Model type i can use for GOMLX , Currently I have Pytorch model , also I can convert it to ONNX model and even if I make some effort probably I can use Tensorflow As well.
So if it possible give me the code Example how can I use it.

**Comment by janpfeifer** [comment]
Sure. So, GoMLX is meant to be a ML framework where you can build your own machine learning models from scratch. The [tutorial](https://gomlx.github.io/gomlx/notebooks/tutorial.html) does that in a very simple model, but the `examples/` subdirectory has many models built and trained from scratch -- I recommend taking a look at [UCI Adult](https://gomlx.github.io/gomlx/notebooks/uci-adult.html).

Now, if you want to use GoMLX simply as a backend to execute an existing model, that is, converting a model from another platform, there is no automated way yet. Usually, it's not hard to manually convert a model, if you understand the code (I converted a the Inceptions models from TensorFlow, and the [Gemma](https://github.com/gomlx/gemma) model from Jax. Under the `gomlx/ml/data` directory there are some file format conversions/donwload.

In parallel, we (with some collaborators) are working right now on something that will convert an ONNX model and build a GoMLX model, see [example here](https://github.com/gomlx/onnx-gomlx/blob/main/onnx-go.ipynb). There are two use cases for that: 

1. Use GoMLX XLA engine (as opposed to `onnxruntime` for ONNX) which may be faster in some cases (and TPU support), we have yet to benchmark.
2. Load an ONNX model in GoMLX and use GoMLX to fine-tune it (it does the auto-differentiation) without requiring python+pytorch, and even export the weights back to the ONNX model. This should be coming the next few days.

**Comment by akshin18** [comment]
Thank you so much for prompt response.

I have tried run this code:
`package main


import (
	"fmt"

	"github.com/janpfeifer/must"

	"github.com/gomlx/gomlx/backends"
	. "github.com/gomlx/gomlx/graph"
	"github.com/gomlx/gomlx/ml/data"
	"github.com/gomlx/onnx-gomlx/onnx"

	_ "github.com/gomlx/gomlx/backends/xla"
)

var (
	_       = Add
	backend = backends.New()
)

func main() {
	var modelPath = data.ReplaceTildeInDir("category_model.onnx") // all-MiniLM-L6-v2
	model := must.M1(onnx.ReadFile(modelPath))
	fmt.Printf("%s\n", model)

}
`


But i got this error: 
`# github.com/gomlx/onnx-gomlx/onnx
../go/pkg/mod/github.com/gomlx/onnx-gomlx@v0.0.0-20241029113425-03b8b422d41a/onnx/graph.go:247:9: undefined: MatMul
../go/pkg/mod/github.com/gomlx/onnx-gomlx@v0.0.0-20241029113425-03b8b422d41a/onnx/materialize.go:19:15: node.ConstantValue undefined (type *graph.Node has no field or method ConstantValue)
../go/pkg/mod/github.com/gomlx/onnx-gomlx@v0.0.0-20241029113425-03b8b422d41a/onnx/ops.go:140:9: undefined: ExpandAxes
../go/pkg/mod/github.com/gomlx/onnx-gomlx@v0.0.0-20241029113425-03b8b422d41a/onnx/ops.go:170:65: node.IsConstantExpression undefined (type *graph.Node has no field or method IsConstantExpression)
# github.com/gomlx/gopjrt/xlabuilder
../go/pkg/mod/github.com/gomlx/gopjrt@v0.4.4/xlabuilder/literal.go:4:10: fatal error: 'gomlx/xlabuilder/literal.h' file not found
    4 | #include <gomlx/xlabuilder/literal.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.`

Note: I use model that has LSTM layer

**Comment by janpfeifer** [comment]
hey @akshin18 , 

While the LSTM layer should be fine to be implemented (or converted to) GoMLX ,  the `onnx-gomlx` repository is not finished yet -- and it is being co-developed with a new release of GoMLX (also in the works) for some things it needs.

You can see how the [example is broken](https://github.com/gomlx/onnx-gomlx/blob/main/onnx-go.ipynb) for the sentence encoder -- not all Ops conversions are done yet.

I was going to suggest to wait for the ONNX converter ... I'm hopeful that by the end of the week or early next week we'll have something that works for most models. I'll keep updating the repository and I'll add the documentation accordingly. Also GoMLX will have a new release.

Btw, you need to install [`gopjrt`](https://github.com/gomlx/gopjrt) manually: it requires a C/C++ library (it uses CGO) installation to talk to XLA/PJRT (the underlying ML engine for accelearators developed by Google). There is a [straight forward script to do that](https://github.com/gomlx/gopjrt?tab=readme-ov-file#installing). Hopefully, at some point I'll package it in Homebrew or Snap.

Let me know if you have any questions.

**Comment by janpfeifer** [comment]
Just an update: https://github.com/gomlx/onnx-gomlx is in a working state already. A sentence encoder has already been fully converted and all seems to work.

Probably there are Ops that are not converted yet, but those are easy. If you try to convert anything that is missing an Op let me know, I'd be happy to fix it.

Currently I'm working on  https://github.com/gomlx/go-huggingface to try to streamline the import of HuggingFace models -- at least those that have an ONNX version.  It's slowly but surely improving :smiley:

**Comment by akshin18** [comment]
Hi @janpfeifer ,
Thank you so much for letting me know , I appreciate  that , will try

**Comment by akshin18** [comment]
I have trie to use next code to try use my model  but it does not work , probably I do not have enough knowledge to understand how to use this library for my model:

package main

// This example illustrates how to use pre-trained vision models.
// model to get the imagenet label for some image.
import (
	"fmt"
	"os"

	"github.com/gomlx/gomlx/backends"
	. "github.com/gomlx/gomlx/graph"
	"github.com/gomlx/gomlx/ml/context"
	"github.com/gomlx/gomlx/ml/data"
	"github.com/gomlx/onnx-gomlx/onnx"
	"github.com/janpfeifer/must"

	_ "github.com/gomlx/gomlx/backends/xla"
)

var (
	_       = Add
	backend = backends.New()
)

func main() {
	var modelPath = data.ReplaceTildeInDir("category_model.onnx") // all-MiniLM-L6-v2
	model := must.M1(onnx.ReadFile(modelPath))
	fmt.Printf("%s\n", model)

	must.M(model.PrintVariables(os.Stdout))
	fmt.Println()
	must.M(model.PrintGraph(os.Stdout))
	fmt.Println()

	// Convert ONNX variables to GoMLX context (which stores variables):
	ctx := context.New()
	must.M(model.VariablesToContext(ctx))
	fmt.Printf("Variables loaded from %q:\n", modelPath)

	gomlxFn := func(ctx *context.Context, x *Node) *Node {
		return model.CallGraph(ctx, x.Graph(), map[string]*Node{"X": x})[0]
	}
	fmt.Println(gomlxFn)
	x := [][]float32{{101, 2070, 2518, 102}}
	fmt.Printf("\tX*A+B=%v\n", context.ExecOnce(backend, ctx, gomlxFn, x).GoStr())
}

**Comment by akshin18** [comment]
Also i tried this code and got that kind of issue :
panic: ONNX node output "" not found as the output of any Op, and not a variable or input either -- could it be a node name, and note a node **output** name ?


My model:
ONNX Model:
        Producer:       pytorch / 2.5.0
        # inputs:       1
                [#0] input: (Int64) [batch_size, 4]
        # outputs:      1
                [#0] output: (Float32) [batch_size, 10]
        # nodes:        13
        # tensors (variables):  6
        # sparse tensors (variables):   0
        Op types:       []string{"Concat", "Constant", "ConstantOfShape", "Gather", "Gemm", "LSTM", "Shape", "Squeeze", "Transpose", "Unsqueeze"}
        IR Version:     6
        Operator Sets:  [v11]
        
        
       
       code:
       ```
       package main

import (
	"fmt"
	"os"

	"github.com/janpfeifer/must"

	"github.com/gomlx/gomlx/backends"
	_ "github.com/gomlx/gomlx/backends/xla"
	. "github.com/gomlx/gomlx/graph"
	"github.com/gomlx/gomlx/ml/context"
	"github.com/gomlx/onnx-gomlx/onnx"
)

func main() {
	// Replace with the path to your LSTM ONNX model
	modelPath := "lstm_model.onnx"

	// Load and parse the ONNX model
	model := must.M1(onnx.ReadFile(modelPath))
	fmt.Printf("%s\n", model)

	// Optionally, print variables and graph for debugging
	must.M(model.PrintVariables(os.Stdout))
	fmt.Println()
	must.M(model.PrintGraph(os.Stdout))
	fmt.Println()

	// Create a new GoMLX context and load model variables into it
	ctx := context.New()
	must.M(model.VariablesToContext(ctx))

	// Prepare input data
	// Assuming batch_size = 2, sequence_length = 5
	inputData := [][]int64{
		{1, 2, 3, 4},
	}

	ou := [][]float32{
		{1, 2, 3, 4},
	}

	// Create a backend (defaults to XLA)
	backend := backends.New()

	// Define the computation function
	gomlxFn := func(ctx *context.Context, x *Node, y *Node) *Node {
		// Map input names to input nodes
		inputsMap := map[string]*Node{
			"input": x,
		}
		// Convert the ONNX model to a GoMLX computation graph
		outputNodes := model.CallGraph(ctx, x.Graph(), inputsMap)
		// Return the first output node (assuming single output)
		return outputNodes[0]
	}

	// Execute the model using the GoMLX context and backend
	outputTensor := context.ExecOnce(backend, ctx, gomlxFn, inputData,ou)

	// Print the output tensor
	fmt.Printf("Model output: %s\n", outputTensor)
}

       ```

**Comment by janpfeifer** [comment]
hi @akshin18 , 

Interesting, thank you for reporting, it helps a lot.

Just yesterday, I manged to get my third model to work -- see the [very last cell in my notebook here](https://github.com/gomlx/onnx-gomlx/blob/main/onnx-go.ipynb).

Also I created [github.com/gomlx/go-huggingface](https://github.com/gomlx/go-huggingface) to download&cache model files from HuggingFace, to make it simpler. Also, I eventually want to add tokenizers support for those doing LLMs.

And you pointed out a couple of new things I need to add support to:

* Unnamed inputs and outputs -- I hadn't notice that ... I'll have to think of something.
* From the Ops listed, I can already say I haven't yet converted the `LSTM` op. 

Is this a model available in HuggingFace that I can try to make it work ? Tomorrow I'll get back to try to expand its support, and it would be nice to have some target models :)

cheers

**Comment by akshin18** [comment]
Hi @janpfeifer ,
Thank you for your reply.
Unfortunately I can not share the model that I use cause it belongs to company, however you can use as a reference this mode: https://www.kaggle.com/code/mehmetlaudatekman/lstm-text-classification-pytorch/notebook

Please do not hesitate to letting me know If i can somehow help you.

Thanks,
Agshin

**Comment by janpfeifer** [comment]
Perfect, I'll use that as model! ETA for having something: mid-next week.

**Comment by janpfeifer** [comment]
hi Agshin,

I just added support for another couple of ONNX models in the list, and I was going to take a look at the LSTM model you pointed to, but there doesn't seem to be a `.onnx` model file I can download in the Kaggle page. Do you have a version of the model not trained (with random initialization) that you could share, so I can use it as test ?

cheers

**Comment by akshin18** [comment]
hi @janpfeifer ,
I do not have exact file , You can use as example this class:
class TextClassificationModel(nn.Module):
    def __init__(self, vocab_size, embedding_dim, hidden_dim, output_size):
        super(TextClassificationModel, self).__init__()
        self.embedding = nn.Embedding(vocab_size, embedding_dim)
        self.lstm = nn.LSTM(embedding_dim, hidden_dim, batch_first=True)
        self.fc = nn.Linear(hidden_dim, output_size)

    def forward(self, x):
        x = self.embedding(x)
        _, (hn, _) = self.lstm(x)
        hn = hn.squeeze(0)
        out = self.fc(hn)
        return out

**Comment by janpfeifer** [comment]
Thanks @akshin18 , but how do you save that to an ONNX file ?

**Comment by akshin18** [comment]
I use that code:`
tokenizer = BertTokenizerFast.from_pretrained("bert-base-uncased")

# Convert to ONNX
dummy_input = torch.randint(0, 30522, (1, 128)).to(device)  # Dummy input tensor
onnx_file_path = "test_.onnx"
torch.onnx.export(
    model,
    dummy_input,
    onnx_file_path,
    input_names=["input"],
    output_names=["output"],
    dynamic_axes={"input": {0: "batch_size", 1: "sequence_length"}, "output": {0: "batch_size"}},
    opset_version=11
)`

**Comment by janpfeifer** [comment]
Thanks again, I'll take a look at it!

**Comment by janpfeifer** [comment]
hi @akshin18 , 

Just an update: I created the LSTM model, and to convert I had to create an LSTM layer in GoMLX as well -- which is a good thing, I've been meaning to do it. 

[Most of the code is already there](https://github.com/gomlx/gomlx/blob/fix/ml/layers/lstm/lstm.go),  but I still need to add tests and stitch together in the converter. Hopefully, I'll get that by tomorrow or Monday if not.

**Comment by janpfeifer** [comment]
LSTM seems to be working now. The new releases of GoMLX v0.15.2 and `onnx-go` v0.1.2 includes the update.

See example of [model in python created here](https://github.com/gomlx/onnx-gomlx/blob/main/onnx-py.ipynb) and the [GoMLX converted version here](https://github.com/gomlx/onnx-gomlx/blob/main/onnx-go.ipynb): both seem to return the same result -- search for the section with LSTM in the title.

**Comment by janpfeifer** [comment]
hi Akshin (@akshin18 ), I'm closing the issue, ok ? But let me know (open another issue or discussion) if you have any questions.
cheers
